### PR TITLE
New walker systems

### PR DESCRIPTION
**Pull request by yangmr04**
# New walker systems

## Description
This PR includes major update to FreeBird.jl, including the implementation of the atomistic and lattice walker systems.

### Live Set types
- `AbstractLiveSet`
  - `LatticeLiveSet`
  - `AtomLiveSet`  

The live sets are vectors of `AbstractWalker` objects, which are used to store the walkers in the system. A live set contains a list of walkers, and a *calculator*, for the calculation of the energy of the system. 

### Walker types
- `AbstractWalker{C}`
  - `LatticeWalker{C}`
  - `AtomWalker{C}`  

where `C` is the number of components in the system, which is an integer. A walker typically has fields for the configuration of the system, the energy, and the iteration number.
  
### Lattice types
- `AbstractLattice`
  - `MLattice{C,G}`

where `C` is the number of components in the system and `G` is the lattice geometry type.  `C` is an integer, while `G` is one of the `LatticeGeometry` types: `SquareLattice`, `TriangularLattice` and `GenericLattice`.
There are also the `SLattice{G}`, which is an alias for `MLattice{1,G}`, and `GLattice{C}`, which is an alias for `MLattice{C,GenericLattice}`. 

### `Potential` and `Hamiltonian` types
For atomistic systems, an interatomic potential is used. We currently only have the standard Lennard-Jones (LJ) potential implemented. 
- `LennardJonesParametersSets`
  - `LJParameters` for a single-component LJ system
  - `CompositeLJParameters` for a multi-component LJ system

For lattice systems, a Hamiltonian is used. 
- `ClassicalHamiltonian`
  - `GenericLatticeHamiltonian`: a generic lattice Hamiltonian that has an on-site interaction term and a list of arbitrary n-th nearest-neighbor interaction terms.
  - `MLatticeHamiltonian`: a multi-component lattice Hamiltonian in the form of a Matrix of `GenericLatticeHamiltonian` objects.

Preview documentation: [FreeBird.jl](https://wexlergroup.github.io/FreeBird.jl/previews/PR63/)

### CODE REVIEW

File: scripts/algorithm-examples.ipynb
Comment (rwexler) [inline scripts/algorithm-examples.ipynb:1]: @yangmr04, what's the point of this file?

File: src/AbstractLiveSets/AbstractLiveSets.jl [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using ..Potentials
+using ..Hamiltonians
+
+export AbstractLiveSet
+export AtomWalkers, LatticeWalkers
+export LJAtomWalkers, LatticeGasWalkers
+
+abstract type AbstractLiveSet end
+
+# Base.eltype(::AbstractLiveSet{T}) where {T} = T
```
Comment (rwexler) [inline src/AbstractLiveSets/AbstractLiveSets.jl:16]: @yangmr04, is this a comment or commented-out code? If it's the latter, should we remove it?

File: src/AbstractLiveSets/LatticeLiveSets.jl [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+# Arguments
+- `walker::LatticeWalker{C}`: The walker to assign energy to.
+- `hamiltonian::LatticeGasHamiltonian`: The Hamiltonian used to calculate the energy.
+- `perturb_energy::Float64=0.0`: The amount of random perturbation to add to the energy.
+
+# Returns
+- `walker::LatticeWalker{C}`: The walker with the assigned energy.
+"""
+function assign_energy!(walker::LatticeWalker{C}, hamiltonian::ClassicalHamiltonian; perturb_energy::Float64=0.0) where C
+    # Assign the energy to the walker, if perturb_energy is non-zero, give all walkers a small random (positive or negative) perturbation
```
Comment (rwexler) [inline src/AbstractLiveSets/LatticeLiveSets.jl:19]:
```suggestion
    # Assign the energy to the walker and, if perturb_energy is non-zero, give all walkers a small random (positive or negative) perturbation
```

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+                                    ) where C
+
+    lattice_vectors = [lattice_constant 0.0 0.0; 0.0 sqrt(3)*lattice_constant 0.0; 0.0 0.0 1.0]
+    lattice_comp, lattice_adsorptions = mlattice_setup(C, basis, supercell_dimensions, components, adsorptions)
+
+    return MLattice{C,TriangularLattice}(lattice_vectors, basis, supercell_dimensions, periodicity, cutoff_radii, lattice_comp, lattice_adsorptions)
+    
+end
+
+"""
```
Comment (yangmr04) [inline src/AbstractWalkers/LatticeWalkers.jl:330]: Oops, docstrings are at the wrong place.

File: src/FreeBirdIO.jl [hunk 4]
```diff
@@ -196,20 +200,23 @@ function convert_walker_to_system(at::AtomWalker)
...
 Append an `AtomWalker` object to a file.
 """
-function append_walker(filename::String, at::AtomWalker)
-    ats = read_walkers(filename)
-    push!(ats, at)
-    write_walkers(filename, ats)
+function append_walker(filename::String, at::AtomWalker{C}) where C
+    # ats = read_walkers(filename)
+    sys = convert_walker_to_system(at)
+    # push!(ats, at)
```
Comment (rwexler) [inline src/FreeBirdIO.jl:217]: @yangmr04, do you want to keep these comments around?

File: src/MonteCarloMoves/MonteCarloMoves.jl [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+export MC_random_walk!, MC_nve_walk!, MC_new_sample!
+export generate_random_new_lattice_sample!
+
+include("helpers.jl")
+
+include("random_walks.jl")
+
+include("nve_walks.jl")
+
+
```
Comment (rwexler) [inline src/MonteCarloMoves/MonteCarloMoves.jl:29]: @yangmr04, should we remove lines 29–37?

File: src/MonteCarloMoves/helpers.jl [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        end
+    end
+    pos = SVector{3,T}(new_pos)
+    return pos
+end
+
+"""
+    mean_sq_displacement(at::AtomWalker, at_orig::AtomWalker)
+
+Calculate the mean squared displacement before and after random walk(s). Note that due to the current implementation of the periodic boundary wrap, this function is not appropriate to use for calculating mean displacements in a propagation.
```
Comment (rwexler) [inline src/MonteCarloMoves/helpers.jl:42]: @yangmr04, is this function worth including if it doesn't compute what most people refer to as the mean squared displacement?

File: src/MonteCarloMoves/nve_walks.jl [hunk 7]
```diff
@@ -178,25 +24,24 @@ Returns:
 - `e_demon::Float64`: The updated energy of the demon.
 """
 function single_atom_demon_walk!(
```
Comment (rwexler) [inline src/MonteCarloMoves/nve_walks.jl:26]: @yangmr04, I'm OK with including the demon walks in the `dev` branch, but let's not release it until we can convince ourselves that it does something more accurately or efficiently than the other sampling algorithms.

File: src/MonteCarloMoves/random_walks.jl [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        samples = sample(unoccupied, occupancy[i], replace=false, ordered=true)
+        for j in samples
+            lattice.components[i][j] = true
+            deleteat!(unoccupied, findfirst(isequal(j), unoccupied))
+        end
+    end
+    return lattice
+end
+
+function generate_random_new_lattice_sample!(lattice::SLattice)
```
Comment (rwexler) [inline src/MonteCarloMoves/random_walks.jl:207]: @yangmr04, many functions are missing docstrings. Does multiple dispatch automatically take care of this, or should we write some docstrings for these, too?

File: src/Potentials.jl [hunk 9]
```diff
@@ -6,6 +6,11 @@ module Potentials
 using Unitful
```
Comment (rwexler) [inline src/Potentials.jl:6]: @yangmr04, I think clear instructions for implementing a new interatomic potential would be helpful to include in the developer guide.

File: src/SamplingSchemes/SamplingSchemes.jl [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+export NestedSamplingParameters, LatticeNestedSamplingParameters
+export sort_by_energy!, nested_sampling_step!
+export nested_sampling_loop!
+export MCRoutine, MCRandomWalkMaxE, MCRandomWalkClone, MCDemonWalk, MixedMCRoutine, MCNewSample
+
+export exact_enumeration, nvt_monte_carlo, wang_landau
+
+abstract type SamplingParameters end
+
+include("NestedSampling.jl")
```
Comment (rwexler) [inline src/SamplingSchemes/SamplingSchemes.jl:28]: @yangmr04, why is the filename formatting scheme different for `NestedSampling.jl` vs. `exact_enumeration.jl`, `nvt_monte_carlo.jl`, etc.

File: src/SamplingSchemes/nvt_monte_carlo.jl [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    push!(configurations, deepcopy(current_lattice))
+
+    for _ in 1:num_steps
+        # Select a random site
+        # site_index = rand(1:length(current_lattice.occupations))
+        
+        # Propose a swap in occupation state (only if it maintains constant N)
+        proposed_lattice = deepcopy(current_lattice)
+        
+        # if proposed_lattice.occupations[site_index] == 1
```
Comment (rwexler) [inline src/SamplingSchemes/nvt_monte_carlo.jl:47]: @yangmr04, lots of commented-out lines here. What should we do with them?

**Review by rwexler — CHANGES_REQUESTED** [review]

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+compute_neighbors(supercell_lattice_vectors::Matrix{Float64}, positions::Matrix{Float64}, cutoff_radii::Tuple{Float64, Float64}, periodicity::Vector{Bool})
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:2]: Good catch! The docstring is outdated!

File: src/SamplingSchemes/SamplingSchemes.jl (hunk 10 above)
Comment (yangmr04) [reply src/SamplingSchemes/SamplingSchemes.jl:28]: Yup, they will be fixed later once they are in the dev branch.

File: src/Potentials.jl (hunk 9 above)
Comment (yangmr04) [reply src/Potentials.jl:6]: Good idea! Aligns with the concept of "building an interface".

File: src/MonteCarloMoves/random_walks.jl (hunk 8 above)
Comment (yangmr04) [reply src/MonteCarloMoves/random_walks.jl:207]: I see. This is an **internal** function that should not be exposed to users. Thus, I don't police myself writing docstings for it. Not a good practice. All functions should have docstings. I will add them.

File: src/MonteCarloMoves/nve_walks.jl (hunk 7 above)
Comment (yangmr04) [reply src/MonteCarloMoves/nve_walks.jl:26]: Agreed! I will move them out to an "experimental" branch later.

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+                                    adsorptions::Union{Vector{Int},Symbol}=:full,
+                                ) where C
+
+    lattice_vectors = [lattice_constant 0.0 0.0; 0.0 lattice_constant 0.0; 0.0 0.0 1.0]
+    lattice_comp, lattice_adsorptions = mlattice_setup(C, basis, supercell_dimensions, components, adsorptions)
+
+    return MLattice{C,SquareLattice}(lattice_vectors, basis, supercell_dimensions, periodicity, cutoff_radii, lattice_comp, lattice_adsorptions)
+end
+
+function MLattice{C,TriangularLattice}(; lattice_constant::Float64=1.0,
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:314]: Yup, this one is missing indeed. Will add.

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+        positions = lattice_positions(lattice_vectors, basis, supercell_dimensions)
+
+        supercell_lattice_vectors = lattice_vectors * Diagonal([supercell_dimensions[1], supercell_dimensions[2], supercell_dimensions[3]])
+        neighbors = compute_neighbors(supercell_lattice_vectors, positions, periodicity, cutoff_radii)
+        
+        return new{C,G}(lattice_vectors, positions, basis, supercell_dimensions, periodicity, cutoff_radii, components, neighbors, adsorptions)
+    end
+end
+
+function split_into_subarrays(arr::AbstractVector, N::Int)
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:234]: Will do.

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+        # Determine the size of the current subarray
+        current_size = base_size + (i <= remainder ? 1 : 0)
+        push!(subarrays, arr[idx:idx + current_size - 1])
+        idx += current_size
+    end
+
+    return subarrays
+end
+
+function mlattice_setup(C::Int,
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:252]: Will do.

File: src/SamplingSchemes/nvt_monte_carlo.jl (hunk 11 above)
Comment (yangmr04) [reply src/SamplingSchemes/nvt_monte_carlo.jl:47]: They will be removed later.

File: src/AnalysisTools.jl [hunk 16]
```diff
@@ -39,8 +39,8 @@ where \$N\$ is the number of walkers and \$i\$ is the iteration number.
 - A vector of \$\\omega\$ factors.
 """
 function ωᵢ(iters::Vector{Int}, n_walkers::Int)
-    ω_i = (1/(n_walkers+1))*(n_walkers/(n_walkers+1)).^iters
-    return ω_i
+    ωi = (1/(n_walkers+1))*(n_walkers/(n_walkers+1)).^iters
+    return ωi
 end
 
 """
```
Comment (junchichen21) [inline src/AnalysisTools.jl:46]: The partition function can return the calculated values when I pass one weight with two energy states or two weights with one energy state. For example, `partition_function(1.0, [1.0], [1.0, 2.0])` returns 0.503214724408055, and `partition_function(1.0, [0.3, 0.7], [1.0])` returns 0.36787944117144233.

But it is incorrect either two energy states can both have weight 1, or one energy state can have two different weights, right?

Comment (yangmr04) [reply src/AnalysisTools.jl:46]: The function does the right thing, but your are using it wrongly. `partition_function(1.0, [1.0], [1.0, 2.0])` case could still make sense, stating that, not only two, but all states shall have the same weight of `1.0`. The `partition_function(1.0, [0.3, 0.7], [1.0])` does not have any physically meaningful interpretation, that you should not have two weights for a single state. While the function allowing to do so, it does not mean that the right thing to do.

Comment (yangmr04) [reply src/AnalysisTools.jl:46]: https://github.com/wexlergroup/FreeBird.jl/blob/92dcf98f58cbae34c3b476f2cbf4cd84303739dd/src/AnalysisTools.jl#L64C1-L69C4
It shows that the function expects `ωi::Vector{Float64},  Ei::Vector{Float64}` that both `ωi` and `Ei` to be vectors, and you pass in a single value (that's wrapped into a vector of a single element, allowing the function to be called). If you try `partition_function(1.0, 1.0, [1.0, 2.0])` or `partition_function(1.0, [0.3, 0.7], 1.0)`, they will error out.

File: src/AbstractLiveSets/AbstractLiveSets.jl (hunk 1 above)
Comment (yangmr04) [reply src/AbstractLiveSets/AbstractLiveSets.jl:16]: Yes, it's the latter. Shall be removed.

File: src/AbstractLiveSets/LatticeLiveSets.jl (hunk 2 above)
Comment (yangmr04) [reply src/AbstractLiveSets/LatticeLiveSets.jl:19]: Done!

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+# Constructor
+```julia
+LatticeWalker(configuration::AbstractLattice; energy=0.0, iter=0)
+```
+Create a new `LatticeWalker` with the given configuration and optional energy and iteration number.
+
+"""
+
+
+# number_of_lattice_components(lattice::SLattice) = 1
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:349]: Done!

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 18]
```diff
@@ -0,0 +1,391 @@
...
+
+const SLattice{G} = MLattice{1,G} # alias for single-component lattices
+
+const GLattice{C} = MLattice{C,GenericLattice} # alias for generic lattices
+
+num_lattice_components(lattice::MLattice{C,G}) where {C,G} = C
+
+function num_sites(lattice::AbstractLattice)
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:357]: Sure!

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+
+function occupied_site_count(MLattice::MLattice{C}) where C
+    occupancy = Array{Int}(undef, C)
+    for i in eachindex(MLattice.components)
+        occupancy[i] = sum(MLattice.components[i])
+    end
+    return occupancy
+end
+
+mutable struct LatticeWalker{C} <: AbstractWalker
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:369]: Okay, this one is actually important.

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 20]
```diff
@@ -0,0 +1,391 @@
...
+    configuration::AbstractLattice
+    energy::typeof(0.0u"eV")
+    iter::Int64
+    function LatticeWalker(configuration::AbstractLattice; energy=0.0u"eV", iter=0)
+        num_comp = num_lattice_components(configuration)
+        return new{num_comp}(configuration, energy, iter)
+    end
+end
+
+function Base.show(io::IO, walker::LatticeWalker)
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:379]: Not needed for `show` methods.

File: src/AbstractWalkers/LatticeWalkers.jl [hunk 21]
```diff
@@ -0,0 +1,391 @@
...
+    println(io, "LatticeWalker(")
+    println(io, "    configuration: ", walker.configuration)
+    println(io, "    energy: ", walker.energy)
+    println(io, "    iter: ", walker.iter, ")")
+end
+
+function Base.show(io::IO, walker::Vector{LatticeWalker})
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/AbstractWalkers/LatticeWalkers.jl:386]: Not needed for `show` methods.

File: src/EnergyEval/EnergyEval.jl [hunk 22]
```diff
@@ -0,0 +1,465 @@
...
+- `pos2::Union{SVector{T},Vector{T}}`: The second position.
+- `at::AbstractSystem`: The abstract system containing boundary conditions and bounding box.
+
+# Returns
+- `d::Float64`: The distance between `pos1` and `pos2` considering periodic boundary conditions.
+
+"""
+function pbc_dist(pos1::Union{SVector{T},Vector{T}},
+                  pos2::Union{SVector{T},Vector{T}},
+                  at::AbstractSystem) where {T}
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/EnergyEval/EnergyEval.jl:36]: No... `AbstractSystem` is something defined in `AtomsBase` where a *system* is referred to an atomistic configuration. Yeah, it's a bit confusing, but it does not cover the lattice thing we implemented.

File: src/EnergyEval/EnergyEval.jl [hunk 23]
```diff
@@ -0,0 +1,465 @@
...
+    end
+    return energy
+end
+
+function inter_component_energy_threaded(at1::AbstractSystem, at2::AbstractSystem, lj::LJParameters)
+    # build pairs of particles
+    pairs = [(i, j) for i in 1:length(at1), j in 1:length(at2)]
+    # @show pairs # DEBUG
+    energy = Array{typeof(0.0u"eV"), 1}(undef, length(pairs))
+    Threads.@threads for k in eachindex(pairs)
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/EnergyEval/EnergyEval.jl:84]: Julia automatically splits the loop into most-equally-lengthened sections, load balance should't be an issue.

File: src/EnergyEval/EnergyEval.jl [hunk 24]
```diff
@@ -0,0 +1,465 @@
...
+end
+
+function interacting_energy(lattice::SLattice, h::MLatticeHamiltonian{C,N,U}) where {C,N,U}
+    ham = h.Hamiltonians[1,1]
+    e_interaction::U = lattice_interaction_energy(lattice.components[1], lattice.neighbors, ham)
+    e_adsorption::U = sum(lattice.components[1] .& lattice.adsorptions) * ham.on_site_interaction
+    return e_interaction + e_adsorption
+end
+
+"""
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/EnergyEval/EnergyEval.jl:438]: Will do! Good idea!

File: src/EnergyEval/EnergyEval.jl [hunk 25]
```diff
@@ -0,0 +1,465 @@
...
+export interacting_energy, frozen_energy
+export sort_components_by_atomic_number
+export split_components
+
+include("helpers.jl")
+
+"""
+    pbc_dist(pos1, pos2, at)
+
+Compute the distance between two positions considering periodic boundary conditions. Currently only works for orthorhombic lattices.
```
[the remark this answers is not in the document]
Comment (yangmr04) [reply src/EnergyEval/EnergyEval.jl:23]: Good idea. Will do it somewhere else. This one here affects performance, checking takes time.

File: src/FreeBirdIO.jl (hunk 4 above)
Comment (yangmr04) [reply src/FreeBirdIO.jl:217]: Gone now!

File: src/MonteCarloMoves/MonteCarloMoves.jl (hunk 5 above)
Comment (yangmr04) [reply src/MonteCarloMoves/MonteCarloMoves.jl:29]: Done!

File: src/MonteCarloMoves/helpers.jl (hunk 6 above)
Comment (yangmr04) [reply src/MonteCarloMoves/helpers.jl:42]: Will move them to a separate file during the tidy-up.

File: scripts/algorithm-examples.ipynb
Comment (yangmr04) [reply scripts/algorithm-examples.ipynb:1]: Please ignore. They will be overhauled later.

**Pull request merged**